sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala

```diff
@@ -438,10 +450,33 @@ abstract class SparkPlanMeta[INPUT <: SparkPlan](plan: INPUT,
   }
 
   private def makeShuffleConsistent(): Unit = {
+    // during query execution when AQE is enabled, the plan could consist of a mixture of
+    // ShuffleExchangeExec nodes for exchanges that have not started executing yet, and
+    // ShuffleQueryStageExec nodes for exchanges that have already started executing. This code
+    // attempts to tag ShuffleExchangeExec nodes for CPU if other exchanges (either
+    // ShuffleExchangeExec or ShuffleQueryStageExec nodes) were also tagged for CPU.
     val exchanges = findShuffleExchanges()
-    if (!exchanges.forall(_.canThisBeReplaced)) {
-      exchanges.foreach(_.willNotWorkOnGpu("other exchanges that feed the same join are" +
-        " on the CPU and GPU hashing is not consistent with the CPU version"))
+    val queryStages = findShuffleQueryStages()
+
+    val consistentExchangeMessage = "other exchanges that feed the same join are" +
+        " on the CPU and GPU hashing is not consistent with the CPU version"
+
+    if (queryStages.isEmpty) {
+      // this is the original logic which works fine when AQE is disabled and also for
+      // initial preparation before query stage creation when AQE is enabled
+      if (!exchanges.forall(_.canThisBeReplaced)) {
+        exchanges.foreach(_.willNotWorkOnGpu(consistentExchangeMessage))
+      }
+    } else if (exchanges.isEmpty) {
+      // we only have query stages, and we cannot do anything to modify them at this point
+      // since they already started to execute
+    } else {
+      // there is a mix of ShuffleExchangeExec and ShuffleQueryStageExec so we need to tag any
+      // ShuffleExchangeExec nodes based on the other nodes
```

Comment:
so if any exchange or query exchange is not on the GPU, all exchanges and query exchanges must fallback to the CPU, that feed to the join? Just making sure that's the end result.

Should that happen also for `queryStages`? E.g. if `exchanges` are on the cpu, but `queryStages` are on the gpu, the following code will just tag the `exchanges` again for the cpu (which seems unnecessary, but I could be misreading it). Wonder if we need to also tell the `queryStages` that they can't work because of the consistent hashing issue.